File: src/rate_of_closure/mesh.py
"""STL clubhead meshes: pure-numpy parser, writer, and head normalization.

Supports the optional photorealistic-clubhead rendering mode. Both STL
flavours are handled with nothing beyond numpy: binary (80-byte header,
uint32 triangle count, 50-byte records) and ASCII (``solid``/``facet``/
``vertex`` text). A user-supplied mesh arrives in arbitrary units and
orientation, so :func:`normalize_head` maps it onto the same canonical
envelope the procedural wireframe uses (AffineDrift frame: x target,
y up, z right):

1. Degenerate (zero-area) triangles are dropped.
2. Axes are permuted by bounding-box extent — largest to z (heel-toe
   width), middle to x (face-to-back depth), smallest to y (crown
   height) — the proportions of every driver head, so the face plate
   ends up facing +x.
3. The bounding box is centered on the origin and scaled uniformly so
   the depth (x extent) equals :data:`HEAD_DEPTH_M`.

The TypeScript twin (``web/src/model/mesh.ts``) implements the same
rules and is parity-tested against the numbers pinned here.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

import numpy as np

from ._contracts import ensure, require

__all__ = [
    "HEAD_DEPTH_M",
    "HeadMesh",
    "load_head_mesh",
    "normalize_head",
    "parse_stl",
    "triangle_normals",
    "write_ascii_stl",
    "write_binary_stl",
]

#: Canonical face-to-back depth of the head envelope [m]; matches the
#: procedural wireframe's ``_BODY_DEPTH`` in the PyQt6 club view.
HEAD_DEPTH_M = 0.11

_BINARY_HEADER_BYTES = 80
_BINARY_RECORD: np.dtype[np.void] = np.dtype(
    [("normal", "<f4", (3,)), ("vertices", "<f4", (3, 3)), ("attr", "<u2")]
)
_ASCII_VERTEX = re.compile(rb"vertex\s+([-+0-9.eE]+)\s+([-+0-9.eE]+)\s+([-+0-9.eE]+)")
_MIN_AREA = 1e-20


@dataclass(frozen=True)
class HeadMesh:
    """A normalized clubhead mesh ready for rendering.

    Attributes:
        triangles: ``(n, 3, 3)`` float64 vertex array in the canonical
            head frame (meters; bounding box centered on the origin).
        normals: ``(n, 3)`` unit normals recomputed from the winding.
    """

    triangles: np.ndarray
    normals: np.ndarray


def parse_stl(data: bytes) -> np.ndarray:
    """Parse STL bytes (binary or ASCII) into ``(n, 3, 3)`` triangles.

    Format detection follows the standard heuristic: a file whose length
    matches the binary record layout is binary (binary headers may start
    with ``solid`` too), otherwise a leading ``solid`` keyword means
    ASCII.
    """
    require(isinstance(data, (bytes, bytearray)), "data must be bytes")
    require(len(data) > 0, "data must not be empty")
    triangles = (
        _parse_binary(bytes(data))
        if _looks_binary(bytes(data))
        else _parse_ascii(bytes(data))
    )
    ensure(triangles.shape[0] > 0, "STL contained no triangles")
    ensure(bool(np.isfinite(triangles).all()), "STL vertices must be finite")
    return triangles


def _looks_binary(data: bytes) -> bool:
    if len(data) < _BINARY_HEADER_BYTES + 4:
        return False
    count = int(
        np.frombuffer(data, dtype="<u4", count=1, offset=_BINARY_HEADER_BYTES)[0]
    )
    return len(data) == _BINARY_HEADER_BYTES + 4 + count * _BINARY_RECORD.itemsize


def _parse_binary(data: bytes) -> np.ndarray:
    count = int(
        np.frombuffer(data, dtype="<u4", count=1, offset=_BINARY_HEADER_BYTES)[0]
    )
    require(count > 0, "binary STL declares zero triangles")
    records = np.frombuffer(
        data, dtype=_BINARY_RECORD, count=count, offset=_BINARY_HEADER_BYTES + 4
    )
    vertices: np.ndarray = records["vertices"].astype(np.float64)
    return vertices


def _parse_ascii(data: bytes) -> np.ndarray:
    require(
        data.lstrip()[:5].lower() == b"solid",
        "not a valid STL: neither binary layout nor ASCII 'solid'",
    )
    matches = _ASCII_VERTEX.findall(data)
    require(len(matches) > 0, "ASCII STL contains no vertex lines")
    require(
        len(matches) % 3 == 0,
        "ASCII STL vertex count must be a multiple of 3",
        len(matches),
    )
    flat = np.array([[float(c) for c in m] for m in matches], dtype=np.float64)
    triangles: np.ndarray = flat.reshape(-1, 3, 3)
    return triangles


def triangle_normals(triangles: np.ndarray) -> np.ndarray:
    """Unit normals from vertex winding, ``(n, 3)``.

    Degenerate triangles get a zero normal rather than NaN so callers
    can filter or shade them safely.
    """
    tris = np.asarray(triangles, dtype=np.float64)
    require(tris.ndim == 3 and tris.shape[1:] == (3, 3), "triangles must be (n, 3, 3)")
    cross = np.cross(tris[:, 1] - tris[:, 0], tris[:, 2] - tris[:, 0])
    length = np.linalg.norm(cross, axis=1, keepdims=True)
    with np.errstate(invalid="ignore", divide="ignore"):
        normals = np.where(length > _MIN_AREA, cross / length, 0.0)
    unit_normals: np.ndarray = np.asarray(normals, dtype=np.float64)
    return unit_normals


def normalize_head(triangles: np.ndarray, depth_m: float = HEAD_DEPTH_M) -> np.ndarray:
    """Map arbitrary triangles onto the canonical head envelope.

    See the module docstring for the three rules (drop degenerates,
    permute axes by extent, center and scale to ``depth_m``).
    """
    require(depth_m > 0.0, "depth_m must be positive", depth_m)
    tris = np.asarray(triangles, dtype=np.float64)
    require(tris.ndim == 3 and tris.shape[1:] == (3, 3), "triangles must be (n, 3, 3)")
    require(bool(np.isfinite(tris).all()), "triangles must be finite")

    areas = np.linalg.norm(
        np.cross(tris[:, 1] - tris[:, 0], tris[:, 2] - tris[:, 0]), axis=1
    )
    tris = tris[areas > _MIN_AREA]
    require(tris.shape[0] > 0, "mesh has no non-degenerate triangles")

    flat = tris.reshape(-1, 3)
    extents = flat.max(axis=0) - flat.min(axis=0)
    require(bool((extents > 0.0).all()), "mesh must have volume on all axes", extents)

    # Stable extent ordering: [smallest, middle, largest] source axes.
    order = np.argsort(extents, kind="stable")
    # middle -> x (depth), smallest -> y (height), largest -> z (width).
    permutation = np.array([order[1], order[0], order[2]])
    tris = tris[:, :, permutation]

    flat = tris.reshape(-1, 3)
    center = (flat.max(axis=0) + flat.min(axis=0)) / 2.0
    scale = depth_m / extents[order[1]]
    normalized = (tris - center) * scale

    span = normalized.reshape(-1, 3).max(axis=0) - normalized.reshape(-1, 3).min(axis=0)
    ensure(bool(np.isclose(span[0], depth_m, rtol=1e-9)), "depth must normalize")
    ensure(span[2] >= span[0] >= span[1], "extent ordering z >= x >= y must hold")
    canonical: np.ndarray = np.asarray(normalized, dtype=np.float64)
    return canonical


def load_head_mesh(path: str | Path, depth_m: float = HEAD_DEPTH_M) -> HeadMesh:
    """Parse an STL file and normalize it into a renderable head mesh."""
    stl_path = Path(path)
    require(stl_path.is_file(), "STL path must be an existing file", str(stl_path))
    triangles = normalize_head(parse_stl(stl_path.read_bytes()), depth_m)
    return HeadMesh(triangles=triangles, normals=triangle_normals(triangles))


def write_binary_stl(triangles: np.ndarray, header: str = "rate_of_closure") -> bytes:
    """Serialize ``(n, 3, 3)`` triangles as a binary STL byte string."""
    tris = np.asarray(triangles, dtype=np.float64)
    require(tris.ndim == 3 and tris.shape[1:] == (3, 3), "triangles must be (n, 3, 3)")
    require(tris.shape[0] > 0, "cannot write an empty STL")
    records = np.zeros(tris.shape[0], dtype=_BINARY_RECORD)
    records["normal"] = triangle_normals(tris).astype(np.float32)
    records["vertices"] = tris.astype(np.float32)
    head = header.encode("ascii", errors="replace")[:_BINARY_HEADER_BYTES]
    head = head.ljust(_BINARY_HEADER_BYTES, b"\0")
    count = np.array([tris.shape[0]], dtype="<u4")
    payload: bytes = head + count.tobytes() + records.tobytes()
    return payload


def write_ascii_stl(triangles: np.ndarray, name: str = "rate_of_closure") -> bytes:
    """Serialize ``(n, 3, 3)`` triangles as an ASCII STL byte string."""
    tris = np.asarray(triangles, dtype=np.float64)
    require(tris.ndim == 3 and tris.shape[1:] == (3, 3), "triangles must be (n, 3, 3)")
    require(tris.shape[0] > 0, "cannot write an empty STL")
    normals = triangle_normals(tris)
    lines = [f"solid {name}"]
    for tri, normal in zip(tris, normals, strict=True):
        lines.append(f"  facet normal {normal[0]:.9e} {normal[1]:.9e} {normal[2]:.9e}")
        lines.append("    outer loop")
        for vertex in tri:
            lines.append(
                f"      vertex {vertex[0]:.9e} {vertex[1]:.9e} {vertex[2]:.9e}"
            )
        lines.append("    endloop")
        lines.append("  endfacet")
    lines.append(f"endsolid {name}")
    return ("\n".join(lines) + "\n").encode("ascii")

File: src/rate_of_closure/web/src/model/mesh.ts
/**
 * STL clubhead meshes — TypeScript twin of `rate_of_closure/mesh.py`.
 *
 * Parses both binary STL (80-byte header, uint32 count, 50-byte
 * records) and ASCII STL (`solid`/`facet`/`vertex`), then normalizes
 * arbitrary meshes onto the canonical head envelope (AffineDrift
 * frame: x target, y up, z right):
 *
 * 1. Degenerate (zero-area) triangles are dropped.
 * 2. Axes are permuted by bounding-box extent — largest to z (heel-toe
 *    width), middle to x (face-to-back depth), smallest to y (crown
 *    height) — so the face plate ends up facing +x.
 * 3. The bounding box is centered on the origin and scaled uniformly
 *    so the depth (x extent) equals `HEAD_DEPTH_M`.
 *
 * Parity-tested in vitest against the numbers pinned by
 * `tests/rate_of_closure/test_mesh.py`.
 */

export type Vec3 = [number, number, number];
export type Triangle = [Vec3, Vec3, Vec3];

export interface HeadMesh {
  /** Triangles in the canonical head frame (meters, bbox centered). */
  triangles: Triangle[];
  /** Unit normals recomputed from the winding (zero if degenerate). */
  normals: Vec3[];
}

/** Canonical face-to-back depth of the head envelope [m]. */
export const HEAD_DEPTH_M = 0.11;

const BINARY_HEADER_BYTES = 80;
const BINARY_RECORD_BYTES = 50;
const BINARY_PREFIX_BYTES = BINARY_HEADER_BYTES + 4;
const MAX_UINT32 = 0xffff_ffff;
const MIN_AREA = 1e-20;
const ASCII_VERTEX = /vertex\s+([-+0-9.eE]+)\s+([-+0-9.eE]+)\s+([-+0-9.eE]+)/g;

function looksBinary(buffer: ArrayBuffer): boolean {
  if (buffer.byteLength < BINARY_HEADER_BYTES + 4) return false;
  const count = new DataView(buffer).getUint32(BINARY_HEADER_BYTES, true);
  return (
    buffer.byteLength ===
    BINARY_HEADER_BYTES + 4 + count * BINARY_RECORD_BYTES
  );
}

function parseBinary(buffer: ArrayBuffer): Triangle[] {
  const view = new DataView(buffer);
  const count = view.getUint32(BINARY_HEADER_BYTES, true);
  if (count === 0) throw new Error("binary STL declares zero triangles");
  const triangles: Triangle[] = [];
  for (let t = 0; t < count; t += 1) {
    // Skip the 12-byte record normal; normals are recomputed.
    const base = BINARY_HEADER_BYTES + 4 + t * BINARY_RECORD_BYTES + 12;
    const tri: Vec3[] = [];
    for (let v = 0; v < 3; v += 1) {
      tri.push([
        view.getFloat32(base + v * 12, true),
        view.getFloat32(base + v * 12 + 4, true),
        view.getFloat32(base + v * 12 + 8, true),
      ]);
    }
    triangles.push(tri as Triangle);
  }
  return triangles;
}

function parseAscii(buffer: ArrayBuffer): Triangle[] {
  const text = new TextDecoder("utf-8", { fatal: false }).decode(buffer);
  if (!text.trimStart().toLowerCase().startsWith("solid")) {
    throw new Error("not a valid STL: neither binary layout nor ASCII 'solid'");
  }
  const vertices: Vec3[] = [];
  for (const match of text.matchAll(ASCII_VERTEX)) {
    vertices.push([Number(match[1]), Number(match[2]), Number(match[3])]);
  }
  if (vertices.length === 0) {
    throw new Error("ASCII STL contains no vertex lines");
  }
  if (vertices.length % 3 !== 0) {
    throw new Error("ASCII STL vertex count must be a multiple of 3");
  }
  const triangles: Triangle[] = [];
  for (let i = 0; i < vertices.length; i += 3) {
    triangles.push([vertices[i], vertices[i + 1], vertices[i + 2]]);
  }
  return triangles;
}

/** Parse STL bytes (binary or ASCII) into triangles. */
export function parseStl(buffer: ArrayBuffer): Triangle[] {
  if (buffer.byteLength === 0) throw new Error("data must not be empty");
  const triangles = looksBinary(buffer)
    ? parseBinary(buffer)
    : parseAscii(buffer);
  if (triangles.length === 0) throw new Error("STL contained no triangles");
  for (const tri of triangles) {
    for (const v of tri) {
      if (!v.every(Number.isFinite)) {
        throw new Error("STL vertices must be finite");
      }
    }
  }
  return triangles;
}

function cross(a: Vec3, b: Vec3): Vec3 {
  return [
    a[1] * b[2] - a[2] * b[1],
    a[2] * b[0] - a[0] * b[2],
    a[0] * b[1] - a[1] * b[0],
  ];
}

function sub(a: Vec3, b: Vec3): Vec3 {
  return [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
}

function triangleCross(tri: Triangle): Vec3 {
  return cross(sub(tri[1], tri[0]), sub(tri[2], tri[0]));
}

/** Unit normals from vertex winding; zero vector for degenerates. */
export function triangleNormals(triangles: Triangle[]): Vec3[] {
  return triangles.map((tri) => {
    const c = triangleCross(tri);
    const len = Math.hypot(...c);
    return len > MIN_AREA
      ? ([c[0] / len, c[1] / len, c[2] / len] as Vec3)
      : ([0, 0, 0] as Vec3);
  });
}

function binaryHeader(header: string): Uint8Array {
  const encoded = new Uint8Array(BINARY_HEADER_BYTES);
  let offset = 0;
  for (const character of header) {
    if (offset >= BINARY_HEADER_BYTES) break;
    const codePoint = character.codePointAt(0);
    encoded[offset] =
      codePoint !== undefined && codePoint <= 0x7f ? codePoint : 0x3f;
    offset += 1;
  }
  return encoded;
}

/** Serialize finite triangles as deterministic little-endian binary STL. */
export function writeBinaryStl(
  triangles: Triangle[],
  header = "rate_of_closure",
): ArrayBuffer {
  if (triangles.length === 0) throw new Error("cannot write an empty STL");
  if (triangles.length > MAX_UINT32) throw new Error("too many STL triangles");
  for (const triangle of triangles) {
    for (const vertex of triangle) {
      if (!vertex.every(Number.isFinite)) {
        throw new Error("STL vertices must be finite");
      }
    }
  }
  const normals = triangleNormals(triangles);
  const buffer = new ArrayBuffer(
    BINARY_PREFIX_BYTES + triangles.length * BINARY_RECORD_BYTES,
  );
  const bytes = new Uint8Array(buffer);
  const view = new DataView(buffer);
  bytes.set(binaryHeader(header));
  view.setUint32(BINARY_HEADER_BYTES, triangles.length, true);
  for (
    let triangleIndex = 0;
    triangleIndex < triangles.length;
    triangleIndex += 1
  ) {
    const recordOffset =
      BINARY_PREFIX_BYTES + triangleIndex * BINARY_RECORD_BYTES;
    const values = [normals[triangleIndex], ...triangles[triangleIndex]];
    for (let vectorIndex = 0; vectorIndex < values.length; vectorIndex += 1) {
      const vectorOffset = recordOffset + vectorIndex * 12;
      const vector = values[vectorIndex];
      for (let axis = 0; axis < 3; axis += 1) {
        view.setFloat32(vectorOffset + axis * 4, vector[axis], true);
      }
    }
  }
  return buffer;
}

function bounds(triangles: Triangle[]): { min: Vec3; max: Vec3 } {
  const min: Vec3 = [Infinity, Infinity, Infinity];
  const max: Vec3 = [-Infinity, -Infinity, -Infinity];
  for (const tri of triangles) {
    for (const v of tri) {
      for (let axis = 0; axis < 3; axis += 1) {
        if (v[axis] < min[axis]) min[axis] = v[axis];
        if (v[axis] > max[axis]) max[axis] = v[axis];
      }
    }
  }
  return { min, max };
}

/** Map arbitrary triangles onto the canonical head envelope. */
export function normalizeHead(
  triangles: Triangle[],
  depthM: number = HEAD_DEPTH_M,
): Triangle[] {
  if (!(depthM > 0)) throw new Error("depthM must be positive");
  const solid = triangles.filter(
    (tri) => Math.hypot(...triangleCross(tri)) > MIN_AREA,
  );
  if (solid.length === 0) {
    throw new Error("mesh has no non-degenerate triangles");
  }
  const { min, max } = bounds(solid);
  const extents: Vec3 = [max[0] - min[0], max[1] - min[1], max[2] - min[2]];
  if (!extents.every((e) => e > 0)) {
    throw new Error("mesh must have volume on all axes");
  }
  // Stable extent ordering: [smallest, middle, largest] source axes.
  const order = [0, 1, 2].sort((a, b) => extents[a] - extents[b] || a - b);
  // middle -> x (depth), smallest -> y (height), largest -> z (width).
  const permutation = [order[1], order[0], order[2]];
  const permuted = solid.map(
    (tri) =>
      tri.map(
        (v) => [v[permutation[0]], v[permutation[1]], v[permutation[2]]] as Vec3,
      ) as Triangle,
  );
  const box = bounds(permuted);
  const center: Vec3 = [
    (box.max[0] + box.min[0]) / 2,
    (box.max[1] + box.min[1]) / 2,
    (box.max[2] + box.min[2]) / 2,
  ];
  const scale = depthM / extents[order[1]];
  return permuted.map(
    (tri) =>
      tri.map(
        (v) =>
          [
            (v[0] - center[0]) * scale,
            (v[1] - center[1]) * scale,
            (v[2] - center[2]) * scale,
          ] as Vec3,
      ) as Triangle,
  );
}

/** Parse and normalize STL bytes into a renderable head mesh. */
export function loadHeadMesh(
  buffer: ArrayBuffer,
  depthM: number = HEAD_DEPTH_M,
): HeadMesh {
  const triangles = normalizeHead(parseStl(buffer), depthM);
  return { triangles, normals: triangleNormals(triangles) };
}
